plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1070 @@
+		klass_rmi->setup = fu_synaptics_rmi_v5_device_setup;
+		klass_device->detach = fu_synaptics_rmi_v5_device_detach;
+		klass_device->write_firmware = fu_synaptics_rmi_v5_device_write_firmware;
+	} else if (priv->f34->function_version == 0x1) {
+		klass_rmi->setup = fu_synaptics_rmi_v6_device_setup;
+		klass_device->detach = fu_synaptics_rmi_v5_device_detach;
+		klass_device->write_firmware = fu_synaptics_rmi_v5_device_write_firmware;
+	} else if (priv->f34->function_version == 0x2) {
+		klass_rmi->setup = fu_synaptics_rmi_v7_device_setup;
+		klass_device->detach = fu_synaptics_rmi_v7_device_detach;
+		klass_device->write_firmware = fu_synaptics_rmi_v7_device_write_firmware;
+	} else {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "f34 function version 0x%02x unsupported",
+			     priv->f34->function_version);
+		return FALSE;
+	}
+
+	/* get Function34_Query0,1 */
+	if (!klass_rmi->setup (self, error)) {
+		g_prefix_error (error, "failed to read f34 queries: ");
+		return FALSE;
+	}
+	bl_ver = g_strdup_printf ("%u.0", priv->flash.bootloader_id[1]);
+	fu_device_set_version_bootloader (device, bl_ver);
+
+	/* in bootloader mode */
+	f01_db = fu_synaptics_rmi_device_read (self, priv->f01->data_base, 0x1, error);
+	if (f01_db == NULL) {
+		g_prefix_error (error, "failed to read the f01 data base: ");
+		return FALSE;
+	}
+	if (f01_db->data[0] & 0x40) {
+		fu_device_add_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	} else {
+		fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_open (FuDevice *device, GError **error)
+{
+	FuSynapticsRmiDevice *self = FU_SYNAPTICS_RMI_DEVICE (device);
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	GUdevDevice *udev_device = fu_udev_device_get_dev (FU_UDEV_DEVICE (device));
+
+g_debug ("OPENING DEVICE %p %s\n\n", device, fu_udev_device_get_device_file (FU_UDEV_DEVICE (device)));
+
+	/* open device */
+	priv->fd = g_open (g_udev_device_get_device_file (udev_device), O_RDWR);
+	if (priv->fd < 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "failed to open %s: %s",
+			     g_udev_device_get_device_file (udev_device),
+			     strerror (errno));
+		return FALSE;
+	}
+	priv->io_channel = fu_io_channel_unix_new (priv->fd);
+
+	/* set up touchpad so we can query it */
+	if (!fu_synaptics_rmi_device_set_mode (self, HID_RMI4_MODE_ATTN_REPORTS, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_close (FuDevice *device, GError **error)
+{
+	FuSynapticsRmiDevice *self = FU_SYNAPTICS_RMI_DEVICE (device);
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GError) error_local = NULL;
+
+	/* turn it back to mouse mode */
+	if (!fu_synaptics_rmi_device_set_mode (self, HID_RMI4_MODE_MOUSE, &error_local)) {
+		/* if just detached for replug, swallow error */
```

Comment:
Do you want that `error_local` to a least come into debugging output though?

Reply:
I guess, added.